### PR TITLE
istioctl: allow pods to be referred to indirectly, using service/<svc> or deployment/<dep>, like kubectl

### PR DESCRIPTION
**Pull request by esnible**
For https://github.com/istio/istio/issues/26080

Instead of knowing the pod name, e.g. `istioctl proxy-status httpbin-6674f664d-fw2hd`, users will be able to imply the pod from other types, similar to how 'kubectl logs' or 'kubectl exec' work.  For example

```
kubectl proxy-status deployment/httpbin
kubectl ps service/httpbin
```

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
lgtm

**Review by richardwxn — APPROVED** [review]

File: istioctl/cmd/proxyconfig.go [hunk 1]
```diff
@@ -299,7 +300,18 @@ func clusterConfigCmd() *cobra.Command {
 			var configWriter *configdump.ConfigWriter
 			var err error
 			if len(args) == 1 {
-				podName, ns := handlers.InferPodInfo(args[0], handlers.HandleNamespace(namespace, defaultNamespace))
+				var client kube.ExtendedClient
+				client, err = kubeClient(kubeconfig, configContext)
```
Comment (shamsher31) [inline istioctl/cmd/proxyconfig.go:304]: Even though each subcommand run independently of each other, they all create an instance of `kubeClient`.
Is there a way we can create `kubeClient` once and share it across subcommands? May be using `PersistentPreRun` ?

File: istioctl/cmd/proxyconfig.go [hunk 2]
```diff
@@ -365,7 +377,18 @@ func listenerConfigCmd() *cobra.Command {
...
+				if err != nil {
+					return fmt.Errorf("failed to create k8s client: %w", err)
+				}
+				var podName, ns string
```
Comment (shamsher31) [inline istioctl/cmd/proxyconfig.go:385]: nit: Maybe move all the declaration in one place?

```go
var (
	configWriter *configdump.ConfigWriter
	err error
	client kube.ExtendedClient
	podName, ns string
)
```

File: istioctl/cmd/proxyconfig.go [hunk 3]
```diff
@@ -579,7 +623,7 @@ func routeConfigCmd() *cobra.Command {
 
 func endpointConfigCmd() *cobra.Command {
 	endpointConfigCmd := &cobra.Command{
-		Use:   "endpoint [<pod-name[.namespace]>]",
+		Use:   "endpoint [<type>/]<name>[.<namespace>]",
```
Comment (shamsher31) [inline istioctl/cmd/proxyconfig.go:626]: Good to have a test for these updates.

**Comment by linsun** [comment]
@esnible this is a great improvement and simplification!

Could we also support the old format for a bit more so that users have time to transition to this new format?

**Comment by esnible** [comment]
> @esnible this is a great improvement and simplification!
> Could we also support the old format for a bit more so that users have time to transition to this new format?

This is completely upwardly compatible, everything that used to work still works.  The only difference is when the user specifies an argument with a `/`, which enables the new logic.

File: istioctl/cmd/proxyconfig.go (hunk 1 above)
Comment (esnible) [reply istioctl/cmd/proxyconfig.go]: Thanks, I decided to put this in `Args` as it is part of argument parsing.

File: istioctl/cmd/proxyconfig.go (hunk 2 above)
Comment (esnible) [reply istioctl/cmd/proxyconfig.go]: As part of the refactor I moved podName and ns (renamed PodNamespace) to their own section, eliminated err and client, and left configWriter alone.

File: istioctl/cmd/proxyconfig.go [hunk 4]
```diff
@@ -766,3 +808,18 @@ func proxyConfig() *cobra.Command {
 
 	return configCmd
 }
+
+func getPodName(podflag string) (string, string, error) {
```
Comment (shamsher31) [inline istioctl/cmd/proxyconfig.go:812]: This looks great. How about moving `getPodName ` into `handlers` and utilize it for `authz`, `dashboard` and `proxy-status`?

**Review by therealmitchconnors — APPROVED** [review]

**Review by linsun — APPROVED** [review]
Thanks @esnible - excellent UX improvement!

github.com/fatih/camelcase is MIT license, LGTM

**Pull request merged**